databricks-sdk-java/src/main/java/com/databricks/sdk/core/oauth/GithubIDTokenSource.java

```
@@ -0,0 +1,110 @@
+    // Validate required configuration
+    if (Strings.isNullOrEmpty(actionsIDTokenRequestURL)) {
+      throw new DatabricksException("Missing ActionsIDTokenRequestURL");
+    }
+    if (Strings.isNullOrEmpty(actionsIDTokenRequestToken)) {
+      throw new DatabricksException("Missing ActionsIDTokenRequestToken");
+    }
+    if (httpClient == null) {
+      throw new DatabricksException("HttpClient cannot be null");
+    }
+
+    String requestUrl = actionsIDTokenRequestURL;
+    if (!Strings.isNullOrEmpty(audience)) {
+      requestUrl = String.format("%s&audience=%s", requestUrl, audience);
+    }
+
+    Request req =
+        new Request("GET", requestUrl)
+            .withHeader("Authorization", "Bearer " + actionsIDTokenRequestToken);
+
+    Response resp;
+    try {
+      resp = httpClient.execute(req);
+    } catch (IOException e) {
+      throw new DatabricksException(
+          "Failed to request ID token from " + requestUrl + ": " + e.getMessage(), e);
+    }
+
+    // Validate response status code
+    if (resp.getStatusCode() != 200) {
+      throw new DatabricksException(
+          "Failed to request ID token: status code "
+              + resp.getStatusCode()
+              + ", response body: "
+              + resp.getBody().toString());
+    }
+
+    // Parse the JSON response
+    ObjectNode jsonResp;
+    try {
+      jsonResp = mapper.readValue(resp.getBody(), ObjectNode.class);
+    } catch (IOException e) {
+      throw new DatabricksException(
+          "Failed to request ID token: corrupted token: " + e.getMessage());
+    }
+
+    // Validate response structure and token value
+    if (!jsonResp.has("value")) {
+      throw new DatabricksException("ID token response missing 'value' field");
+    }
+
+    String tokenValue = jsonResp.get("value").textValue();
+    if (Strings.isNullOrEmpty(tokenValue)) {
+      throw new DatabricksException("Received empty ID token from GitHub Actions");
+    }
+
+    return new IDToken(tokenValue);
```

IDToken constructor can throw IIlegalArgumentException. Since we are not catching it, let's add that in the doccomment.

---

This will never happen though given the if statement in line 104. 

In almost all cases, we should model optionality using `Optional` and not `null`. It is thus reasonable to assume that a `String` is never null. In that regard, I'm wondering if we should simplify the constructor of the IdToken class so that it does not throw an exception. 

We could do something like this if we really want to be defensive:

```java
public class IDToken {
  // The string value of the ID Token
  private final String value;

  /**
   * Constructs an IDToken with a value.
   *
   * @param value The ID Token string.
   */
  public IDToken(String value) {
    this.value = Strings.isNullOrEmpty(value) ? "" : value;
  }

  /**
   * Returns the value of the ID Token.
   *
   * @return The string representation of the ID Token.
   */
  public String getValue() {
    return value;
  }
}
```

---

>It would thus be the responsibility of the functions receiving ID tokens to decide whether they are OK with an empty token or not.

Empty Token will always be invalid, right? I can't imagine a scenario where an empty token is valid. So, It would be better to catch at construction. Also, failing fast is a recommended approach in JAVA.

> In almost all cases, it is preferable to model optionality using Optional and not null. That way, we can consider that non-optional objects are never null, except if specifically indicated as nullable.

Agreed, this validation is doing the same by ensuring value can never be null.

---

> Empty Token will always be invalid, right? I can't imagine a scenario where an empty token is valid.

It's a decision we have to make. I'm personally comfortable with allowing the jwt string to be empty and let the client decide. Though, I'm happy to keep it as-is too.

> failing fast is a recommended approach in JAVA.

Fair but that is true for most things in life, not really specific to this case. One could argue that this is exactly what line 104 is about. 

I don't think listing the runtime exceptions of the underlying functions is a scalable solution though. If we're going to keep the check in the constructor, then I would prefer something like this:

```java
    try {
      String tokenValue = jsonResp.get("value").textValue();
      return new IdToken(tokenValue);
    } catch (IllegalArgumentException e) {
      throw new IllegalArgumentException("...");
    };
```

In the same line of thought, should we move the validation at the beginning of `getIdToken` to the constructor?

---

Following the previous PR, @hectorcast-db suggested moving the parameter validation from the constructor to the getToken() method for consistency with other providers, which only validate required fields when a token is actually requested. I adapted this change for the DatabricksOAuthTokenSource class, which accepts an IDTokenSource as input. If validation were to fail during construction, it would prevent the instantiation of DatabricksOAuthTokenSource altogether. As this may fail client construction, I decided to defer validation to getIDToken() instead as this would be called inside getToken() of the wrapper class. @hectorcast-db Let me know if my understanding is correct :)

---

Thanks for the context @emmyzhou-db.

>  If validation were to fail during construction, it would prevent the instantiation of DatabricksOAuthTokenSource altogether. 

Is there a scenario in which it would fail at construction but succeed later?

---

See my message above. Same reasoning applies.

---

I'm not sure I agree with that; `IdTokenSources` are configured at construction.

---

From offline discussion:

- IdTokenSources should ideally be validated at creation. The reason we cannot do that at the moment is because we are not creating them within the `configure` function of their corresponding `CredentialsProvider`. 
- We will add a TODO in the code to make that clear and address that in a follow-up PR. For now, let's keep validating in the `getIdToken()` method.